src/clean_air/models.py

```diff
@@ -27,21 +34,879 @@ class ContactDetails:
+    the timedelta class doesn't support weeks, months, or years. Presumably to avoid issues arising from ambiguity
+    caused by different calendars.
+
+    We've also taken the decision that Durations will be naive: they won't try to take into account differences caused
+    by calendars.
+
+    For arithmetic between Duration instances, we've decided that because:
+    * python's timedelta assumes days are always 24 hours long, we will also assume that
+    * the length of months vary, we can't assume 1 month == 30 days.
+      * hence, we won't convert from days->months or months->days, so
+        * P20D + P20D would result in P40D and not P1M10D.
+        * P1M would not be converted to P30D
+      * we won't allow decimal fractions for months or years (since decimal fractions in years means we have to deal
+        with fractions of a month)
+
+    For the addition/subtraction of Durations from datetimes, because we'd be adding/removing time from a known
+    reference point, we would be able to take into account the length of months, daylight savings time, etc
+
+    Considering the python timedelta class, it treats days as being 24 hours long, so we view this as acceptable too.
+    """
+    # This class is intended to be read-only. Similar to timedelta, most operations that would cause a change should
+    # return a new instance. Here are some things that we could implement, but haven't needed as yet:
+    # TODO: .replace() method to create a copy of the Duration with the given fields replaced with the supplied values
+    # TODO: .remove() method to create a copy of the Duration with the given fields removed
+
+    # Constants used during various field conversions
+    _MONTHS_IN_YEAR = 12
+    _DAYS_IN_WEEK = 7
+    _MICROSECONDS_IN_SECOND = 1000000
+    _SECONDS_IN_MINUTE = 60
+    _SECONDS_IN_HOUR = 60 * _SECONDS_IN_MINUTE
+    _SECONDS_IN_DAY = 24 * _SECONDS_IN_HOUR
+    _SECONDS_IN_WEEK = 7 * _SECONDS_IN_DAY
+    _SECONDS_IN_YEAR = 365 * _SECONDS_IN_DAY  # Common year, rather than Gregorian year (and definitely not Julian!)
+
+    # These regexes are for parsing and extracting the values from ISO 8601 duration strings
+    # I feel I should apologise for what follows, but also assure it could be a lot worse...
+    # Regex for matching a single field, parameterised to accept the match group name and unit character
+    _FLOAT_REGEX_FMT = r"(?:(?P<{}>\d+(?:\.\d+)?){})?"
+    # Regex for parsing and extracting most of an ISO 8601 duration string
+    _PARSER = re.compile("".join([
+        "^P",
+        _FLOAT_REGEX_FMT.format("years", "Y"),
+        _FLOAT_REGEX_FMT.format("months", "M"),
+        _FLOAT_REGEX_FMT.format("days", "D"),
+        "(?:T",  # T is required if any of the following fields are present
+        _FLOAT_REGEX_FMT.format("hours", "H"),
+        _FLOAT_REGEX_FMT.format("minutes", "M"),
+        _FLOAT_REGEX_FMT.format("seconds", "S"),
+        ")?",
```

Comment:
yea, tell me about it...

Reply:
I really don't like all this WKT stuff (which this ISO format string is effectively an example of) - it seems exessively hard to parse on a computer while a person can pull all the information out quickly and easily. I spent quite a while trying in vain to find a system to parse the CRS / VRS / TRS format strings used in the EDR spec (I got as far as abstract syntax trees before deciding that really they just recreated the problem in a whole load of Python objects rather than a single string) before deciding I didn't need them anyway. I just want a system that pulls all the important stuff out of the WKT and stores in in JSON or something...